Web Extensions: Defining functions for the Bookmark IDL and Add API stubs for Bookmarks

akutira-umich opened the pull request:
#### 983a51abfffb41f145632d8fd784060511578298
<pre>
Web Extensions: Defining functions for the Bookmark IDL and Add API stubs for Bookmarks
<a href="https://rdar.apple.com/153338800">rdar://153338800</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294479">https://bugs.webkit.org/show_bug.cgi?id=294479</a>

Reviewed by Brian Weinstein.

Put 11 bookmark functions defined in the .IDL, .h, and cocoa.mm file.
Tells JS layer an error saying &quot;unimplemented&quot;. Test files included that tests
that the functions aren&apos;t undefined and incorrectly typed arguments vend errors.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getChildren):
(WebKit::WebExtensionAPIBookmarks::getSubTree):
(WebKit::WebExtensionAPIBookmarks::move):
(WebKit::WebExtensionAPIBookmarks::remove):
(WebKit::WebExtensionAPIBookmarks::removeTree):
(WebKit::WebExtensionAPIBookmarks::update):
(WebKit::WebExtensionAPIBookmarks::createBookmarks): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl:

Canonical link: <a href="https://commits.webkit.org/296455@main">https://commits.webkit.org/296455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bd3ea02d6cd328307cf7f65d4e8f92f1c765b1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18676 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113799 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58981 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28941 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36806 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82476 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97808 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62913 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15947 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58513 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15998 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116920 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91500 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91303 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23262 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36201 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13962 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35546 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41079 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35256 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38602 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36934 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/925bae4aff3dad8982ffe3f10ab67c87de1145e0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46936

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+#if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)
+
+namespace WebKit {
+
```
Nit: Extra blank line here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    callback->reportError(@"unimplemented");
+}
+
+void WebExtensionAPIBookmarks::getChildren(NSString *id, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
`id` is a protected symbol in Objective-C, we should probably use a different name here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    callback->reportError(@"unimplemented");
+}
+
+void WebExtensionAPIBookmarks::getSubTree(NSString *id, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
Ditto about not using `id` here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    callback->reportError(@"unimplemented");
+}
+
+void WebExtensionAPIBookmarks::move(NSString *id, NSDictionary *destination, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
Ditto everywhere else it's used below.

b-weinstein replied:
Maybe `bookmarkIdentifier` or `bookmarkID`?

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+    callback->reportError(@"unimplemented");
+}
+
+void WebExtensionAPIBookmarks::search(NSObject *query, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
What format does this object take?

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -35,6 +35,30 @@ namespace WebKit {
+
+    void search(NSObject *query, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString);
+
+    void update(NSString *id, NSDictionary *changes, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString);
```
Super-nit: I think instead of doing a blank line between each of these, they could be grouped by type of method.

`create`

`get*`

`remove*`

`search`
`update`
`move`.

akutira-umich commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,95 @@
+namespace WebKit {
+
+
+void WebExtensionAPIBookmarks::createBookmarks(NSDictionary *bookmark, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```
I named it bookmark because that was the name listed in the MDN docs: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/bookmarks/create. It's only creating one bookmark, I can change that to make the name more clear.

xeenon reviewed (changes requested):
Requesting changes based on @b-weinstein's review comments.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b89c15cf672f0bb63e49515a74efd4a674bcf25e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46936

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a1c29ce36dd9ba8d791549ae7c8aff32696e6dd3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46936

b-weinstein commented:
`that tests that the functions aren't undefined and none NSDictionary parameters are correct.`

Since not all the methods take NSDictionary parameters, this line could be a bit more open, something like: `and incorrectly typed arguments vend errors`.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fd79aa003ac5a59b4a3d15b8491ad1cf788dce4e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46936

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5bd3ea02d6cd328307cf7f65d4e8f92f1c765b1d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46936

akutira-umich committed:
Web Extensions: Defining functions for the Bookmark IDL and Add API stubs for Bookmarks
rdar://153338800
https://bugs.webkit.org/show_bug.cgi?id=294479

Reviewed by Brian Weinstein.

Put 11 bookmark functions defined in the .IDL, .h, and cocoa.mm file.
Tells JS layer an error saying "unimplemented". Test files included that tests
that the functions aren't undefined and incorrectly typed arguments vend errors.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getChildren):
(WebKit::WebExtensionAPIBookmarks::getSubTree):
(WebKit::WebExtensionAPIBookmarks::move):
(WebKit::WebExtensionAPIBookmarks::remove):
(WebKit::WebExtensionAPIBookmarks::removeTree):
(WebKit::WebExtensionAPIBookmarks::update):
(WebKit::WebExtensionAPIBookmarks::createBookmarks): Deleted.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl:

Canonical link: https://commits.webkit.org/296455@main

webkit-commit-queue commented:
Committed 296455@main (983a51abfffb): <https://commits.webkit.org/296455@main>

Reviewed commits have been landed. Closing PR #46936 and removing active labels.
